File: script/make/protocol/js/BuffProtocol.js
export default class BuffProtocol {
    static encode(textEncoder, view, offset, protocol, data) {
        switch (protocol) {
            case 11801: {
                return new DataView(view.buffer.slice(0, offset));
            }
            default: throw("unknown protocol define: " + protocol)
        }
    }

    static decode(textDecoder, view, offset, protocol) {
        switch (protocol) {
            case 11801: {
                // Buff列表
                const list = [];
                let listLength = view.getUint16(offset, false);
                offset = offset + 2;
                while (--listLength >= 0) {
                    // Buff
                    // BuffID
                    const buffId = view.getUint32(offset, false);
                    offset = offset + 4;
                    // 结束时间
                    const expireTime = view.getUint32(offset, false);
                    offset = offset + 4;
                    // 叠加数量
                    const overlap = view.getUint16(offset, false);
                    offset = offset + 2;
                    // object
                    const buff = {buffId, expireTime, overlap};
                    // add
                    list.push(buff);
                }
                return {list};
            }
            case 11802: {
                // Buff列表
                const list = [];
                let listLength = view.getUint16(offset, false);
                offset = offset + 2;
                while (--listLength >= 0) {
                    // Buff
                    // BuffID
                    const buffId = view.getUint32(offset, false);
                    offset = offset + 4;
                    // object
                    const buff = {buffId};
                    // add
                    list.push(buff);
                }
                return {list};
            }
            default: throw("unknown protocol define: " + protocol)
        }
    }
}

File: script/make/protocol/lua/BuffProtocol.lua
BuffProtocol = {}

function BuffProtocol.encode(offset, protocol, data)
    if protocol == 11801 then
        local table = {}

        return table
    else
        error(string.format('unknown protocol define: %d', protocol))
    end
end

function BuffProtocol.decode(offset, protocol, bytes)
    if protocol == 11801 then
        -- Buff列表
        local data = {}
        local dataLength = string.unpack(">I2", bytes, offset)
        offset = offset + 2
        for dataIndex = 1, dataLength do
            -- 
            -- BuffID
            local dataDataBuffId = string.unpack(">I4", bytes, offset)
            offset = offset + 4
            -- 结束时间
            local dataDataExpireTime = string.unpack(">I4", bytes, offset)
            offset = offset + 4
            -- 叠加数量
            local dataDataOverlap = string.unpack(">I2", bytes, offset)
            offset = offset + 2
            -- object
            local dataData = {buffId = dataDataBuffId, expireTime = dataDataExpireTime, overlap = dataDataOverlap}
            data[dataIndex] = dataData
        end
        return data
    elseif protocol == 11802 then
        -- BuffID列表
        local data = {}
        local dataLength = string.unpack(">I2", bytes, offset)
        offset = offset + 2
        for dataIndex = 1, dataLength do
            -- BuffID
            local dataData = string.unpack(">I4", bytes, offset)
            offset = offset + 4
            data[dataIndex] = dataData
        end
        return data
    else
        error(string.format('unknown protocol define: %d', protocol))
    end
end